File: Tests/Conformance/StorageFacadePrototype05823.c
#include <assert.h>
#include <stdint.h>
#include <string.h>

#include <r4os/r4os.h>

static const uint8_t file_bytes[] = "hello";
static const uint8_t config_bytes[] =
    "\xEF\xBB\xBF" "R4S_FORMAT=1\r\nSCHEMA=TEST\r\nTITLE=Configured\r\nCOUNT=27\r\n";
static uint8_t written[64];
static uint32_t written_length;
static uint8_t stream_active;
static uint8_t stream_aborted;
static uint64_t registry_value;
static uint16_t registry_type;

static int32_t fake_write(const uint8_t *bytes, uint32_t length) {
    uint32_t count = length < sizeof(written) ? length : (uint32_t)sizeof(written);
    memcpy(written, bytes, count);
    written_length = count;
    return (int32_t)count;
}

static void fake_putc(uint8_t byte) {
    if (written_length < sizeof(written)) written[written_length++] = byte;
}

static int32_t fake_file_read(const uint8_t *path, uint8_t *out, uint32_t capacity) {
    const uint8_t *source = strstr((const char *)path, "CONFIG") != 0 ? config_bytes : file_bytes;
    uint32_t length = source == config_bytes ? (uint32_t)sizeof(config_bytes) - 1u : (uint32_t)sizeof(file_bytes) - 1u;
    uint32_t count = length < capacity ? length : capacity;
    memcpy(out, source, count);
    return (int32_t)count;
}

static int32_t fake_file_read_at(const uint8_t *path, uint32_t offset, uint8_t *out, uint32_t capacity) {
    (void)path;
    uint32_t length = (uint32_t)sizeof(file_bytes) - 1u;
    if (offset >= length) return 0;
    uint32_t count = length - offset < capacity ? length - offset : capacity;
    memcpy(out, file_bytes + offset, count);
    return (int32_t)count;
}

static int32_t fake_file_write(const uint8_t *path, const uint8_t *bytes, uint32_t length) {
    (void)path;
    return fake_write(bytes, length);
}

static int32_t fake_file_append(const uint8_t *path, const uint8_t *bytes, uint32_t length) {
    (void)path;
    uint32_t available = (uint32_t)sizeof(written) - written_length;
    uint32_t count = length < available ? length : available;
    memcpy(written + written_length, bytes, count);
    written_length += count;
    return (int32_t)count;
}

static int32_t fake_file_info(const uint8_t *path, R4FileInfo *info) {
    if (strstr((const char *)path, "MISSING") != 0) return 0;
    *info = (R4FileInfo){0}; info->exists = 1u; info->size = sizeof(file_bytes) - 1u; return 1;
}

static int32_t fake_file_delete(const uint8_t *path) {
    return strstr((const char *)path, "MISSING") != 0 ? 0 : 1;
}

static int32_t fake_file_rename(const uint8_t *source, const uint8_t *target) {
    (void)source;
    (void)target;
    return 1;
}

static int32_t fake_dir_create(const uint8_t *path) {
    (void)path;
    return 1;
}

static int32_t fake_dir_entry(const uint8_t *path, uint32_t index, uint8_t *out, uint32_t capacity) {
    (void)path;
    const char *entry = index == 2u ? "C:\\ONE.TXT" : "C:\\SUB";
    if (index > 3u) return -5;
    uint32_t length = (uint32_t)strlen(entry);
    if (length + 1u > capacity) return -4;
    memcpy(out, entry, length + 1u);
    return index == 3u ? 1 : 0;
}

static int32_t fake_stream_begin(const uint8_t *path, uint32_t flags) {
    (void)path;
    (void)flags;
    stream_active = 1u;
    stream_aborted = 0u;
    written_length = 0u;
    return R4OS_FILE_STREAM_RESULT_OK;
}

static int32_t fake_stream_write(const uint8_t *path, uint64_t offset, const uint8_t *bytes, uint32_t length, uint32_t flags) {
    (void)flags;
    if (!stream_active || offset != written_length) return R4OS_FILE_STREAM_ERROR_OFFSET_MISMATCH;
    return fake_file_append(path, bytes, length);
}

static int32_t fake_stream_finish(const uint8_t *path, uint64_t expected, uint32_t flags) {
    (void)path;
    (void)flags;
    if (!stream_active || expected != written_length) return R4OS_FILE_STREAM_ERROR_SIZE_MISMATCH;
    stream_active = 0u;
    return R4OS_FILE_STREAM_RESULT_OK;
}

static int32_t fake_stream_abort(const uint8_t *path) {
    (void)path;
    stream_active = 0u;
    stream_aborted = 1u;
    return R4OS_FILE_STREAM_RESULT_OK;
}

static int32_t fake_registry_get(const uint8_t *key, const uint8_t *name, R4RegistryValueInfo *info, uint8_t *out, uint32_t capacity) {
    (void)key;
    if (strcmp((const char *)name, "MISSING") == 0) return R4OS_REGISTRY_API_RESULT_VALUE_NOT_FOUND;
    *info = (R4RegistryValueInfo){0};
    if (strcmp((const char *)name, "TEXT") == 0) {
        if (capacity < 2u) return R4OS_REGISTRY_API_RESULT_BUFFER_TOO_SMALL;
        info->value_type = R4OS_REGISTRY_VALUE_TYPE_STRING; info->data_len = 2u; memcpy(out, "OK", 2u); return 2;
    }
    if (strcmp((const char *)name, "WIDE") == 0) {
        if (capacity < 8u) return R4OS_REGISTRY_API_RESULT_BUFFER_TOO_SMALL;
        info->value_type = R4OS_REGISTRY_VALUE_TYPE_U64; info->data_len = 8u; for (uint32_t i = 0; i < 8u; ++i) out[i] = (uint8_t)(0x0102030405060708ull >> (i * 8u)); return 8;
    }
    if (strcmp((const char *)name, "ENABLED") == 0) {
        if (capacity < 1u) return R4OS_REGISTRY_API_RESULT_BUFFER_TOO_SMALL;
        info->value_type = R4OS_REGISTRY_VALUE_TYPE_BOOL; info->data_len = 1u; out[0] = 1u; return 1;
    }
    if (capacity < 4u) return R4OS_REGISTRY_API_RESULT_BUFFER_TOO_SMALL;
    info->value_type = R4OS_REGISTRY_VALUE_TYPE_U32; info->data_len = 4u;
    out[0] = 0x78u; out[1] = 0x56u; out[2] = 0x34u; out[3] = 0x12u; return 4;
}

static int32_t fake_registry_set(const uint8_t *key, const uint8_t *name, uint16_t type, const uint8_t *bytes, uint32_t length) {
    (void)key;
    (void)name;
    registry_type = type; registry_value = 0u;
    if (length > 8u) return R4OS_REGISTRY_API_RESULT_INVALID;
    for (uint32_t i = 0; i < length; ++i) registry_value |= (uint64_t)bytes[i] << (i * 8u);
    return R4OS_REGISTRY_API_RESULT_OK;
}

static int32_t fake_registry_delete(const uint8_t *key, const uint8_t *name) {
    (void)key; return strcmp((const char *)name, "MISSING") == 0 ? R4OS_REGISTRY_API_RESULT_VALUE_NOT_FOUND : R4OS_REGISTRY_API_RESULT_OK;
}

static R4XStartR4Sys make_table(int full) {
    R4XStartR4Sys table = {0};
    table.write = (uint64_t)(uintptr_t)&fake_write;
    table.putc = (uint64_t)(uintptr_t)&fake_putc;
    if (!full) return table;
    table.file_read = (uint64_t)(uintptr_t)&fake_file_read;
    table.file_write = (uint64_t)(uintptr_t)&fake_file_write;
    table.file_read_at = (uint64_t)(uintptr_t)&fake_file_read_at;
    table.file_append = (uint64_t)(uintptr_t)&fake_file_append;
    table.file_info = (uint64_t)(uintptr_t)&fake_file_info;
    table.file_delete = (uint64_t)(uintptr_t)&fake_file_delete;
    table.file_rename = (uint64_t)(uintptr_t)&fake_file_rename;
    table.dir_create = (uint64_t)(uintptr_t)&fake_dir_create;
    table.dir_delete = (uint64_t)(uintptr_t)&fake_file_delete;
    table.dir_entry = (uint64_t)(uintptr_t)&fake_dir_entry;
    table.file_stream_begin = (uint64_t)(uintptr_t)&fake_stream_begin;
    table.file_stream_write = (uint64_t)(uintptr_t)&fake_stream_write;
    table.file_stream_finish = (uint64_t)(uintptr_t)&fake_stream_finish;
    table.file_stream_abort = (uint64_t)(uintptr_t)&fake_stream_abort;
    table.registry_get_value = (uint64_t)(uintptr_t)&fake_registry_get;
    table.registry_set_value = (uint64_t)(uintptr_t)&fake_registry_set;
    table.registry_delete_value = (uint64_t)(uintptr_t)&fake_registry_delete;
    return table;
}

int main(void) {
    R4XStartR4Sys table = make_table(1);
    R4App app = {0};
    app.system.table = &table;
    R4Files files = r4_app_files(&app);
    R4FilePath path = {0};
    assert(r4_file_path((const uint8_t *)"C:/TEMP/FILE.TXT", 16u, &path) == R4_PATH_OK);

    uint8_t buffer[128] = {0};
    R4Transfer transfer = r4_files_read(&files, &path, buffer, sizeof(buffer));
    assert(transfer.state == R4_TRANSFER_BYTES && transfer.bytes == 5u);
    assert(r4_files_write(&files, &path, (const uint8_t *)"A", 1u).state == R4_TRANSFER_BYTES);
    assert(r4_files_append(&files, &path, (const uint8_t *)"B", 1u).state == R4_TRANSFER_BYTES);
    assert(written_length == 2u && memcmp(written, "AB", 2u) == 0);
    assert(r4_files_info(&files, &path).state == R4_FILE_INFO_VALUE);

    R4FilePath directory = {0};
    assert(r4_file_path((const uint8_t *)"C:\\", 3u, &directory) == R4_PATH_OK);
    R4DirectoryIterator iterator = r4_files_iterate(files, &directory);
    assert(r4_directory_next(&iterator).state == R4_DIRECTORY_ENTRY);
    assert(r4_directory_next(&iterator).is_directory == 1u);
    assert(r4_directory_next(&iterator).state == R4_DIRECTORY_END);
    assert(r4_directory_next(&iterator).state == R4_DIRECTORY_END);

    R4StreamReader reader = r4_files_stream_reader(files, &path);
    assert(r4_stream_reader_read(&reader, buffer, 2u).bytes == 2u);
    assert(r4_stream_reader_read(&reader, buffer, sizeof(buffer)).bytes == 3u);
    assert(r4_stream_reader_read(&reader, buffer, sizeof(buffer)).state == R4_TRANSFER_END);

    R4StreamWriter writer = {0};
    assert(r4_files_stream_writer(files, &path, R4OS_FILE_STREAM_OPEN_REPLACE, &writer).state == R4_OPERATION_OK);
    assert(r4_stream_writer_write(&writer, (const uint8_t *)"stream", 6u).state == R4_OPERATION_OK);
    assert(r4_stream_writer_finish(&writer).state == R4_OPERATION_OK);
    assert(r4_files_stream_writer(files, &path, R4OS_FILE_STREAM_OPEN_REPLACE, &writer).state == R4_OPERATION_OK);
    assert(r4_stream_writer_abort(&writer).state == R4_OPERATION_OK && stream_aborted);

    R4Registry registry = r4_app_registry(&app);
    R4RegistryPath key = {0};
    assert(r4_registry_path((const uint8_t *)"HKCU/Software/Test", 18u, &key) == R4_PATH_OK);
    R4RegistryRead value = r4_registry_get(&registry, &key, "COUNT", buffer, sizeof(buffer));
    uint32_t decoded = 0u;
    assert(value.state == R4_REGISTRY_VALUE && r4_registry_value_u32(&value.value, &decoded));
    assert(decoded == 0x12345678u);
    assert(r4_registry_set_u32(&registry, &key, "COUNT", 99u).state == R4_OPERATION_OK);
    assert(registry_type == R4OS_REGISTRY_VALUE_TYPE_U32 && registry_value == 99u);
    value = r4_registry_get(&registry, &key, "WIDE", buffer, sizeof(buffer));
    uint64_t wide = 0u; assert(r4_registry_value_u64(&value.value, &wide) && wide == 0x0102030405060708ull);
    value = r4_registry_get(&registry, &key, "ENABLED", buffer, sizeof(buffer));
    int enabled = 0; assert(r4_registry_value_bool(&value.value, &enabled) && enabled);
    value = r4_registry_get(&registry, &key, "TEXT", buffer, sizeof(buffer));
    R4TextView text = {0}; assert(r4_registry_value_string(&value.value, &text) && text.len == 2u);
    assert(r4_registry_set_u64(&registry, &key, "WIDE", wide).state == R4_OPERATION_OK);
    assert(registry_type == R4OS_REGISTRY_VALUE_TYPE_U64 && registry_value == wide);
    assert(r4_registry_set_bool(&registry, &key, "ENABLED", 1).state == R4_OPERATION_OK);
    assert(registry_type == R4OS_REGISTRY_VALUE_TYPE_BOOL && registry_value == 1u);
    assert(r4_registry_delete(&registry, &key, "COUNT").state == R4_OPERATION_OK);
    assert(r4_registry_delete(&registry, &key, "MISSING").state == R4_OPERATION_MISSING);

    R4XStartR4Sys partial = make_table(0);
    app.system.table = &partial;
    files = r4_app_files(&app);
    registry = r4_app_registry(&app);
    assert(!r4_files_available(&files));
    assert(r4_files_read(&files, &path, buffer, sizeof(buffer)).raw_code == R4OS_ERR_NO_FN);
    assert(!r4_registry_available(&registry));

    uint8_t long_path[R4OS_FILE_PATH_MAX_BYTES + 2u];
    memset(long_path, 'A', sizeof(long_path));
    assert(r4_file_path(long_path, sizeof(long_path), &path) == R4_PATH_TOO_LONG);
    return 0;
}

File: Tests/Conformance/StorageFacadePrototype05823.zig
const std = @import("std");
const r4os = @import("r4os");

const file_bytes = "hello";
const config_bytes = "\xEF\xBB\xBFR4S_FORMAT=1\r\nSCHEMA=TEST\r\nTITLE=Configured\r\nCOUNT=27\r\n";
var written: [64]u8 = .{0} ** 64;
var written_len: usize = 0;
var stream_active = false;
var stream_aborted = false;
var registry_set_value: u32 = 0;
var registry_set_wide: u64 = 0;
var registry_set_type: u16 = 0;
var storage_close_result: i32 = 0;
fn fakeStorageEnd(claim: u64, flags: u32) callconv(.c) i32 {
    std.debug.assert(claim == 17 and flags == 0);
    return storage_close_result;
}
fn fakeStorageBegin(_: *const r4os.abi.StorageTarget, out: *u64) callconv(.c) i32 {
    out.* = 17;
    return 0;
}

test "physical storage checks append-only availability and consuming close outcomes" {
    var table = makeSys(true);
    table.size = r4os.abi.r4xstart_r4sys_size;
    var imports: [1]r4os.abi.R4XStartImport = undefined;
    var context: r4os.abi.R4XStartContext = undefined;
    const app = try makeApp(&table, &imports, &context);
    const sys = app.system();
    const storage = r4os.storage.Context{ .sys = &sys };
    var use: u64 = 123;
    try std.testing.expect(!storage.available());
    try std.testing.expectEqual(@as(i32, 0), storage.transferUseBegin("E:\\FILE", &use));
    try std.testing.expectEqual(@as(u64, 0), use);
    table.size = @sizeOf(r4os.abi.R4XStartR4Sys);
    table.storage_claim_begin = @intFromPtr(&fakeStorageBegin);
    try std.testing.expectEqual(r4os.abi.storage_error_unsupported, storage.transferUseBegin("E:\\FILE", &use));
    const target = r4os.storage.Context.wholeDevice(.{});
    var bytes: [512]u8 = undefined;
    try std.testing.expectEqual(r4os.abi.storage_error_invalid, storage.read(&target, 0, bytes[0..511]));
    try std.testing.expectEqual(@as(i32, 0), storage.claimBegin(&target, &use));
    table.storage_claim_end = @intFromPtr(&fakeStorageEnd);
    storage_close_result = r4os.abi.storage_error_busy;
    try std.testing.expectEqual(storage_close_result, storage.claimEnd(&use, false));
    try std.testing.expectEqual(@as(u64, 17), use);
    storage_close_result = r4os.abi.storage_error_io;
    try std.testing.expectEqual(storage_close_result, storage.claimEnd(&use, false));
    try std.testing.expectEqual(@as(u64, 0), use);
}

fn pathSpan(path: [*:0]const u8) []const u8 {
    return std.mem.span(path);
}

fn copyOut(source: []const u8, out: [*]u8, capacity: u32) i32 {
    const count = @min(source.len, @as(usize, capacity));
    if (count != 0) @memcpy(out[0..count], source[0..count]);
    return @intCast(count);
}

fn fakeWrite(data: [*]const u8, len: u32) callconv(.c) i32 {
    const count = @min(@as(usize, len), written.len);
    @memcpy(written[0..count], data[0..count]);
    written_len = count;
    return @intCast(count);
}

fn fakePutc(ch: u8) callconv(.c) void {
    if (written_len < written.len) {
        written[written_len] = ch;
        written_len += 1;
    }
}

fn fakeFileRead(path: [*:0]const u8, out: [*]u8, capacity: u32) callconv(.c) i32 {
    return copyOut(if (std.mem.indexOf(u8, pathSpan(path), "CONFIG") != null) config_bytes else file_bytes, out, capacity);
}

fn fakeFileReadAt(path: [*:0]const u8, offset: u32, out: [*]u8, capacity: u32) callconv(.c) i32 {
    _ = path;
    if (offset >= file_bytes.len) return 0;
    return copyOut(file_bytes[offset..], out, capacity);
}

fn fakeFileWrite(path: [*:0]const u8, data: [*]const u8, len: u32) callconv(.c) i32 {
    _ = path;
    return fakeWrite(data, len);
}

fn fakeFileAppend(path: [*:0]const u8, data: [*]const u8, len: u32) callconv(.c) i32 {
    _ = path;
    const count = @min(@as(usize, len), written.len - written_len);
    @memcpy(written[written_len..][0..count], data[0..count]);
    written_len += count;
    return @intCast(count);
}

fn fakeFileInfo(path: [*:0]const u8, out: *r4os.abi.FileInfo) callconv(.c) i32 {
    if (std.mem.indexOf(u8, pathSpan(path), "MISSING") != null) return 0;
    out.* = .{ .exists = 1, .size = file_bytes.len };
    return 1;
}

fn fakeDelete(path: [*:0]const u8) callconv(.c) i32 {
    return if (std.mem.indexOf(u8, pathSpan(path), "MISSING") != null) 0 else 1;
}

fn fakeRename(old: [*:0]const u8, new: [*:0]const u8) callconv(.c) i32 {
    _ = old;
    _ = new;
    return 1;
}

fn fakeDirCreate(path: [*:0]const u8) callconv(.c) i32 {
    _ = path;
    return 1;
}

fn fakeDirEntry(path: [*:0]const u8, index: u32, out: [*]u8, capacity: u32) callconv(.c) i32 {
    _ = path;
    if (index == 2) {
        _ = copyOut("C:\\ONE.TXT\x00", out, capacity);
        return 0;
    }
    if (index == 3) {
        _ = copyOut("C:\\SUB\x00", out, capacity);
        return 1;
    }
    return -5;
}

fn fakeStreamBegin(path: [*:0]const u8, flags: u32) callconv(.c) i32 {
    _ = path;
    _ = flags;
    stream_active = true;
    stream_aborted = false;
    written_len = 0;
    return r4os.abi.file_stream_result_ok;
}

fn fakeStreamWrite(path: [*:0]const u8, offset: u64, data: [*]const u8, len: u32, flags: u32) callconv(.c) i32 {
    _ = flags;
    if (!stream_active or offset != written_len) return r4os.abi.file_stream_error_offset_mismatch;
    return fakeFileAppend(path, data, len);
}

fn fakeStreamFinish(path: [*:0]const u8, expected: u64, flags: u32) callconv(.c) i32 {
    _ = path;
    _ = flags;
    if (!stream_active or expected != written_len) return r4os.abi.file_stream_error_size_mismatch;
    stream_active = false;
    return r4os.abi.file_stream_result_ok;
}

fn fakeStreamAbort(path: [*:0]const u8) callconv(.c) i32 {
    _ = path;
    stream_active = false;
    stream_aborted = true;
    return r4os.abi.file_stream_result_ok;
}

fn fakeRegistryGet(key: [*:0]const u8, name: [*:0]const u8, info: *r4os.abi.RegistryValueInfo, out: [*]u8, capacity: u32) callconv(.c) i32 {
    _ = key;
    const value_name = pathSpan(name);
    if (std.mem.eql(u8, value_name, "MISSING")) return r4os.abi.registry_api_result_value_not_found;
    if (std.mem.eql(u8, value_name, "TEXT")) {
        if (capacity < 2) return r4os.abi.registry_api_result_buffer_too_small;
        info.* = .{ .value_type = r4os.abi.registry_value_type_string, .data_len = 2 };
        out[0] = 'O';
        out[1] = 'K';
        return 2;
    }
    if (std.mem.eql(u8, value_name, "WIDE")) {
        if (capacity < 8) return r4os.abi.registry_api_result_buffer_too_small;
        info.* = .{ .value_type = r4os.abi.registry_value_type_u64, .data_len = 8 };
        const value: u64 = 0x0102030405060708;
        for (0..8) |index| out[index] = @truncate(value >> @intCast(index * 8));
        return 8;
    }
    if (std.mem.eql(u8, value_name, "ENABLED")) {
        if (capacity < 1) return r4os.abi.registry_api_result_buffer_too_small;
        info.* = .{ .value_type = r4os.abi.registry_value_type_bool, .data_len = 1 };
        out[0] = 1;
        return 1;
    }
    if (capacity < 4) return r4os.abi.registry_api_result_buffer_too_small;
    info.* = .{ .value_type = r4os.abi.registry_value_type_u32, .data_len = 4 };
    out[0] = 0x78;
    out[1] = 0x56;
    out[2] = 0x34;
    out[3] = 0x12;
    return 4;
}

fn fakeRegistrySet(key: [*:0]const u8, name: [*:0]const u8, value_type: u16, data: [*]const u8, len: u32) callconv(.c) i32 {
    _ = key;
    _ = name;
    if (len > 8) return r4os.abi.registry_api_result_invalid;
    registry_set_type = value_type;
    registry_set_wide = 0;
    for (0..len) |index| registry_set_wide |= @as(u64, data[index]) << @intCast(index * 8);
    registry_set_value = @truncate(registry_set_wide);
    return r4os.abi.registry_api_result_ok;
}

fn makeSys(full: bool) r4os.abi.R4XStartR4Sys {
    var table: r4os.abi.R4XStartR4Sys = .{};
    table.write = @intFromPtr(&fakeWrite);
    table.putc = @intFromPtr(&fakePutc);
    if (!full) return table;
    table.file_read = @intFromPtr(&fakeFileRead);
    table.file_write = @intFromPtr(&fakeFileWrite);
    table.file_read_at = @intFromPtr(&fakeFileReadAt);
    table.file_append = @intFromPtr(&fakeFileAppend);
    table.file_info = @intFromPtr(&fakeFileInfo);
    table.file_delete = @intFromPtr(&fakeDelete);
    table.file_rename = @intFromPtr(&fakeRename);
    table.dir_create = @intFromPtr(&fakeDirCreate);
    table.dir_delete = @intFromPtr(&fakeDelete);
    table.dir_entry = @intFromPtr(&fakeDirEntry);
    table.file_stream_begin = @intFromPtr(&fakeStreamBegin);
    table.file_stream_write = @intFromPtr(&fakeStreamWrite);
    table.file_stream_finish = @intFromPtr(&fakeStreamFinish);
    table.file_stream_abort = @intFromPtr(&fakeStreamAbort);
    table.registry_get_value = @intFromPtr(&fakeRegistryGet);
    table.registry_set_value = @intFromPtr(&fakeRegistrySet);
    return table;
}

fn makeApp(table: *r4os.abi.R4XStartR4Sys, imports: *[1]r4os.abi.R4XStartImport, context: *r4os.abi.R4XStartContext) !r4os.App {
    imports.* = .{
        .{ .group_id = @intFromEnum(r4os.abi.R4LGroup.r4sys), .flags = r4os.abi.r4xstart_import_flag_group_interface, .table = @intFromPtr(table) },
    };
    context.* = .{
        .flags = r4os.abi.r4xstart_flag_imports_valid,
        .imports = @intFromPtr(imports),
        .import_count = imports.len,
    };
    return switch (r4os.App.init(context, .console)) {
        .value => |app| app,
        .failure => error.AppInit,
    };
}

test "files directory streams and registry use caller owned state" {
    var table = makeSys(true);
    var imports: [1]r4os.abi.R4XStartImport = undefined;
    var context: r4os.abi.R4XStartContext = undefined;
    var app = try makeApp(&table, &imports, &context);
    var files = app.files() orelse return error.FilesMissing;
    var file_path = try r4os.FilePath.parse("C:/TEMP/FILE.TXT");
    var out: [16]u8 = undefined;
    switch (files.read(file_path.asZ(), out[0..])) {
        .bytes => |count| try std.testing.expectEqualStrings(file_bytes, out[0..count]),
        else => return error.ReadFailed,
    }
    switch (files.write(file_path.asZ(), "A")) {
        .bytes => |count| try std.testing.expectEqual(@as(u32, 1), count),
        else => return error.WriteFailed,
    }
    switch (files.append(file_path.asZ(), "B")) {
        .bytes => |count| try std.testing.expectEqual(@as(u32, 1), count),
        else => return error.AppendFailed,
    }
    try std.testing.expectEqualStrings("AB", written[0..written_len]);

    var directory = try r4os.FilePath.parse("C:\\");
    var iterator = files.iterate(directory.asZ());
    var entry_buffer: [128]u8 = undefined;
    switch (iterator.next(entry_buffer[0..])) {
        .entry => |entry| try std.testing.expect(entry.kind == .file),
        else => return error.EntryMissing,
    }
    switch (iterator.next(entry_buffer[0..])) {
        .entry => |entry| try std.testing.expect(entry.kind == .directory),
        else => return error.EntryMissing,
    }
    try std.testing.expect(iterator.next(entry_buffer[0..]) == .end);
    try std.testing.expect(iterator.next(entry_buffer[0..]) == .end);

    var reader = files.streamReader(file_path.asZ());
    switch (reader.read(out[0..2])) {
        .bytes => |count| try std.testing.expectEqual(@as(u32, 2), count),
        else => return error.StreamRead,
    }
    switch (reader.read(out[0..])) {
        .bytes => |count| try std.testing.expectEqual(@as(u32, 3), count),
        else => return error.StreamRead,
    }
    try std.testing.expect(reader.read(out[0..]) == .end);

    var writer = switch (files.streamWriter(file_path.asZ(), r4os.abi.file_stream_open_replace)) {
        .writer => |value| value,
        .failure => return error.StreamOpen,
    };
    try std.testing.expect(writer.write("stream") == .ok);
    try std.testing.expect(writer.finish() == .ok);
    var abort_writer = switch (files.streamWriter(file_path.asZ(), r4os.abi.file_stream_open_replace)) {
        .writer => |value| value,
        .failure => return error.StreamOpen,
    };
    try std.testing.expect(abort_writer.abort() == .ok and stream_aborted);

    var registry = app.registry() orelse return error.RegistryMissing;
    var key = try r4os.RegistryPath.parse("HKCU/Software/Test");
    switch (registry.get(&key, "COUNT", out[0..])) {
        .value => |value| try std.testing.expectEqual(@as(?u32, 0x12345678), value.asU32()),
        else => return error.RegistryRead,
    }
    try std.testing.expect(registry.setU32(&key, "COUNT", 99) == .ok);
    try std.testing.expectEqual(@as(u32, 99), registry_set_value);
    switch (registry.get(&key, "WIDE", out[0..])) {
        .value => |value| try std.testing.expectEqual(@as(?u64, 0x0102030405060708), value.asU64()),
        else => return error.RegistryRead,
    }
    switch (registry.get(&key, "ENABLED", out[0..])) {
        .value => |value| try std.testing.expectEqual(@as(?bool, true), value.asBool()),
        else => return error.RegistryRead,
    }
    switch (registry.get(&key, "TEXT", out[0..])) {
        .value => |value| try std.testing.expectEqualStrings("OK", value.asString() orelse return error.RegistryType),
        else => return error.RegistryRead,
    }
    try std.testing.expect(registry.setU64(&key, "WIDE", 0x0102030405060708) == .ok);
    try std.testing.expectEqual(r4os.abi.registry_value_type_u64, registry_set_type);
    try std.testing.expectEqual(@as(u64, 0x0102030405060708), registry_set_wide);
    try std.testing.expect(registry.setBool(&key, "ENABLED", true) == .ok);
    try std.testing.expectEqual(r4os.abi.registry_value_type_bool, registry_set_type);
}

test "missing capability and long paths fail before hidden fallback" {
    var table = makeSys(false);
    var imports: [1]r4os.abi.R4XStartImport = undefined;
    var context: r4os.abi.R4XStartContext = undefined;
    var app = try makeApp(&table, &imports, &context);
    try std.testing.expect(app.files() == null);
    try std.testing.expect(app.registry() == null);
    var too_long: [r4os.path.file_path_max + 1]u8 = .{'A'} ** (r4os.path.file_path_max + 1);
    try std.testing.expectError(error.TooLong, r4os.FilePath.parse(too_long[0..]));
}
